## Differential Expression Analysis of RNA-seq data

For the HTML version of the workflow containing the input/output information and the embedded plots see: [DESeq2.html](https://htmlpreview.github.io/?https://github.com/nshanian/DESeq2/blob/main/DESeq2.html)

## Introduction

The RNA-seq data used in this workflow was generated as part of a study that investigated the regulatory role of short-chain fatty acids (SCFAs) propionate and butyrate in the context of colorectal cancer (CRC). Briefly, epigenomic and transcriptomic profiles were compared in SCFA-treated and untreated CRC cells, normal cells and in mouse intestines, with the goal of gaining insight into changes in accessibility and expression of CRC-relevant genes. 

The data set used in this workflow comes from propionyl RNA-seq in CRC cells. The data demonstrates a significant upregulation in genes controlling development and anatomical structure morphogenesis, and a downregulation in genes controlling cell cycle, cell division, and chromatin organization following treatment with sodium propionate.        

Once the data matrix and metadata file are ready to be loaded into R, `BiocManager`, `pheatmap`, `ggplot2`, `DESeq2`, `EnhancedVolcano` packages will first have to be installed.  

Once all the required files have been prepared and the dependencies installed, _DESeq2_ will be used to identify differentially expressed genes.

## Goals

This workflow will perform the following analyses:

* Compare correlations between replicates and conditions
* Compare expression levels across samples
* Perform principal component analysis (PCA)
* Display fold change of normalized counts between two conditions
* Identify differentially expressed genes 

## Setup

This R Markdown (.Rmd) file will provide a walk-through of exploratory data analysis followed by differential expression analysis of bulk RNA-seq data. Each `Rmd` file has an initial `setup` chunk that applies some settings to all of the code chunks.  

```{r setup, echo = F}
# when you "knit" this file, do you want the resulting PDF to print the code in each chunk (TRUE = yes)?
knitr::opts_chunk$set(echo = TRUE)

################################################################################
# set your working directory
####CHANGE THIS TO THE APPROPRIATE PATH
knitr::opts_knit$set(root.dir = '~/Desktop/Data/RNA-seq/')
################################################################################
# note that outside of an Rmd code chunk, use `setwd()` to set the working directory in R
```

As in a regular R script in RStudio, a single line of code can be run with Command-Enter (Mac OS) or Ctrl-Enter (Windows). Whole chunks of code can be run with Command(/Ctrl) + Shift + Enter **or** by clicking the green ">" button in the top-right corner of the chunk. Alternatively, these options can also be implemented by selecting lines of code and choosing the desired option from the drop-down menu in the "Run" tab, in the top-right corner of the of Source section of RStudio.

To comment or uncomment a series of lines, highlight the lines and use Command(/Ctrl) + Shift + C.  

R packages `DESeq2`, `pheatmap` and `EnhancedVolcano` are required for this workflow.  
They can be installed using the following commands:  
```{r install packages, eval=F}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq2")
install.packages("pheatmap")
install.packages("EnhancedVolcano")

# If install.packages("packagename") command fails in newer versions of R uncomment and run the commands:
# if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("packagename")
```

Once installed, load the packages.
```{r load libraries}
library(pheatmap)
library(DESeq2)
```

## Exploratory Data Analysis

Before analyzing for differential expression, it's a good idea to first visualize the data and perform quality control checks. The first step is to look at `counts_final.txt`, which contains the log-transformed gene expression generated by `rsem`.

Load the matrix of log-transformed gene expression counts. Verify using the `head` and `dim` commands that the counts matrix loaded correctly.

```{r read counts}
log_counts = read.table("~/Desktop/Data/RNA-seq/counts_final.txt")
head(log_counts)
```

First, check the size of the data matrix using `dim(log_counts)`.  
```{r q genes}
dim(log_counts)
```

If there's a significant difference between gene expression in treated and untreated conditions, then samples of similar types should be more correlated with one another compared with samples of different types. This can be visualized by making a scatter plot in base R or `ggplot2` package.
```{r scatterplot}
# Compare counts in two untreated control samples
plot(log_counts[,1], log_counts[,3])
abline(0,1)

# # Alternative approach using ggplot:
library(ggplot2)
ggplot(log_counts, aes(x=Cnt1, y=Cnt3)) +
  geom_point() +
  geom_abline() +
  theme_classic()
```

Compute the correlation of gene expression between two control samples.
```{r correlation}
cor(log_counts[,1], log_counts[,3])
```

Now, modify the previous code to make a scatterplot comparing the expression in a treated sample to the untreated sample. (You can select the _n_-th column of your `log_counts` matrix by typing `log_counts[,n]`).
```{r treated v untreated}
head(log_counts)
plot(log_counts[,'Cnt1'], log_counts[,'Prop1'])
# or plot(log_counts[,1], log_counts[,8])
abline(0,1)
cor(log_counts[,'Cnt1'], log_counts[,'Prop1'])
```

The correlation between replicates from different conditions should be lower than that of replicates from the same condition. 

It is also useful to compute the pairwise correlation between all the different samples at once:
```{r all correlations}
cor(log_counts)
```

To visualize this matrix as a heatmap:
```{r heatmap}
log_cor <- cor(log_counts)
diag(log_cor) <- NA
pheatmap(log_cor, show_rownames = TRUE)
```

Note: The treated (Prop) and untreated (Cnt) conditions should separate into distinct clusters.

Even if the two groups don't separate into perfect clusters, that's not always a problem. But plotting heatmaps can help identify batch effects in the data.

## Detecting batch effects

In this experiment, similar samples cluster together, and we observe two well-defined groups from the two conditions. Often there are batch effects, which can make some of the samples from the same sequencing run appear more similar than they really are.

[Batch effects](http://www.molmine.com/magma/global_analysis/batch_effect.html) are unwanted patterns in sequencing data that arise because of the way the samples have been handled before or during sequencing. For example, samples were analyzed in different sequencing runs (or even different lanes within the same sequencing run) batch effects may be visible.

In some cases, batch effects can completely overpower the signal of interest. Luckily, there are several precautionary steps that can minimize the risk of false interpretations due to batch effects.

### Loading covariates

It's always a good idea to note which samples were sequenced on which sequencing run, as well as other important metadata. An example file `covariates.txt` that contains such information is provided in this repository. Load it into R:
```{r load covariates}
covariates = read.table("~/Desktop/Data/RNA-seq/covariates.txt")
```

Take a look at the covariates.
```{r}
head(covariates)
table(covariates$condition)
```
### Principal Component Analysis

A method called [Principal Component Analysis](https://en.wikipedia.org/wiki/Principal_component_analysis), or PCA, is a type of [Multivariate Analysis](https://web.stanford.edu/class/bios221/book/07-chap.html#the-pca-workflow) used to detect and visualize batch effects in sequencing data. When batch effects or other effects due to experimental covariates are present in the data, they can cause major variation in many of the samples' gene expression. However, it's possible that the same set of genes will be affected similarly across all the affected samples. 

PCA is primarily an exploratory technique that produces maps that show the relations between variables and between observations in a useful way. It uses geometrical projections that take points in higher-dimensional spaces and projects them down onto lower dimensions. PCA is an unsupervised learning technique because, as in clustering, it treats all variables as having the same status. It finds a mathematical model for an underlying structure for all the variables. 

At a high level, PCA finds sources of variation across samples, and then outputs values called _principal components_ that in this example represent the sets of varying genes and to what extent they are present in each sample. If batch effects are present, they will often be detectable by looking at the first few principal components of the gene expression matrix.

The principal components in this RNA-seq data set can be visualized by running the chunk below. Dots represent samples that are color-coded by condition or status.

```{r pca color by condition}
pca = prcomp(t(log_counts)) # "features"/measurements (genes in this case) need to be in the columns

# A simple function for color-coding points
color_code <- function(x, vals, colors)
{
  i = which(x == vals)
  return(colors[i])
}

# Plot PCA using ggplot:
head(pca$x)
pcs = data.frame(pca$x)
pcs$status = as.character(covariates$status[match(rownames(pca$x), as.character(covariates$samples))])
ggplot(pcs, aes(x=PC1, y=PC2, color=status)) +
   geom_point() +
   theme_classic() +
   scale_colour_manual(values=c(Cnt='blue',Prop='red'))
```

Removing batch effects is an important first step in any gene expression analysis. It can be particularly tricky when the experimental covariates are not recorded; however, software tools such as [PEER](https://www.ncbi.nlm.nih.gov/pubmed/22343431) can be used to detect unknown covariates. **Note:** PEER is incompatible with newer versions of R (3.6 and later). 

## Differential expression

If no batch effects are observed then proceed with differential expression analysis. See documentation for the R package [DESeq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8) that quantifies the extent to which gene expression changes between conditions.

As input, DESeq2 takes non-normalized counts values. It then statistically models the probability of differential expression, which involves placing greater confidence on differentially expressed genes for which a greater number of counts are observed across samples. 

Load the final expression counts file `counts_final.txt` again to make sure it is in the global environment.

```{r adjusted counts}
counts = read.table("~/Desktop/Data/RNA-seq/counts_final.txt")
```

Run DESeq using the following commands:
```{r deseq}
rownames(covariates) = covariates$samples
head(covariates)
dds <- DESeqDataSetFromMatrix(countData <- counts,
                              colData <- covariates,
                              design = ~ factor(status, levels=c("Cnt","Prop")))
dds <- DESeq(dds, betaPrior=FALSE)
res = results(dds) # this gets the differentially expressed genes for a contrast
## because we set the levels in the design contrast, this is equivalent to:
# results(dds, contrast=c('status','Cnt','Prop'))
```

DESeq produces an _MA plot_ that shows the estimated log fold change in gene expression in an experimental group vs control, plotted against the mean expression of the gene across samples. Points plotted in blue are significantly differentially expressed between samples. Those plotted in grey are not.
```{r MA plot}
plotMA(dds)
```

The MA plot shows the relationship between the mean expression and the magnitude of log2 fold change. Each blue dot represents a gene that has been identified as differentially expressed between the two conditions. 

In this example the treated group was compared to the untreated group, expressed as Cnt vs Prop or Cnt/Prop ratio. Positive fold change represents increase in expression, or upregulation, in the treated group vs control, while negative fold change represents a decrease in expression, or downregulation in the treated vs control. 

The significance of the calculated fold change for each point or gene is then tested and a corresponding p-value generated. 

DESeq provides two different p-values, `pvalue` and `padj`. When searching for differential expression across many genes, it is always advisable to use `padj`, which corrects for [multiple testing](http://www.stat.berkeley.edu/~mgoldman/Section0402.pdf).
This is analogous to the False-Discovery Rate (FDR) adjusted q-values.

This will determine how many genes are differentially expresssed using the adjusted p-values.  

```{r n degs}
sum(res$padj < 0.05, na.rm=TRUE)
```

Generate a list of top differentially expressed genes.

```{r order res}
head(res[order(res$padj),])
```
Generate a volcano plot to visualize the relationship between fold change and adj p-value.

```{r volcano plot}
library(EnhancedVolcano)

# Check the length of row names
length_of_row_names <- length(rownames(res))

# Print the length
print(length_of_row_names)

# Create the volcano plot only if row names are not empty
if (length_of_row_names > 0) {
  volcano_plot <- EnhancedVolcano(res,
                                   lab = rep("", nrow(res)),  # Leave labels blank
                                   x = "log2FoldChange",
                                   y = "padj",
                                   xlim = c(-4, 4),
                                   title = NULL,         # Remove title
                                   subtitle = NULL,      # Remove subtitle
                                   caption = NULL,       # Remove caption
                                   pCutoff = 0.05,
                                   FCcutoff = 2,
                                   legendLabels = NULL,  # Remove legend
                                   legendPos = "none",   # Remove legend position
                                   legendLabSize = 0     # Remove legend label size
  )
  
  # Remove grid lines and other theme adjustments
  volcano_plot + theme(panel.grid = element_blank(),
                       panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(),
                       axis.line = element_line(colour = "black"))
} else {
  print("Error: Row names are empty.")
}
```

Similar to an MA plot, a volcano plot provides a useful visualization between genes that show positive vs genes that show negative fold change in expression. In this example Prop/Cnt or treated/untreated fold change is positive.

```{r relative fold change}
# Check if 'log2FoldChange' and 'padj' columns exist in the results object
if ("log2FoldChange" %in% colnames(res) && "padj" %in% colnames(res)) {
  # Set the significance threshold
  significance_threshold <- 0.05

  # Count the number of significant peaks with positive fold change
  significant_positive_fold_change <- sum(res$log2FoldChange > 0 & res$padj < significance_threshold, na.rm = TRUE)

  # Count the number of significant peaks with negative fold change
  significant_negative_fold_change <- sum(res$log2FoldChange < 0 & res$padj < significance_threshold, na.rm = TRUE)

  # Print or use the counts as needed
  cat("Number of significant peaks with positive fold change:", significant_positive_fold_change, "\n")
  cat("Number of significant peaks with negative fold change:", significant_negative_fold_change, "\n")
} else {
  print("Error: 'log2FoldChange' or 'padj' columns not found in the results object.")
}
```

Based on this analysis, as a result of treating colorectal cancer (CRC) cells with sodium propionate (NaPr), 2754 genes underwent positive fold change in expression (upregulation), and 2329 genes underwent negative fold change in expression (downregulation) compared to the control, with a significance cutoff of padj < 0.05.

According to DESeq, the top differentially expressed genes is ENSG00000205426. Make a boxplot to verify this.
```{r single gene}
condition <- c(rep("Cnt", 3), rep("Prop", 3))

boxplot(as.numeric(log_counts[rownames(log_counts) == "ENSG00000205426", ]) ~ condition, 
        ylab = "log counts",  # Set y-axis label
        main = "ENSG00000205426",  # Set plot title
        col = c("blue", "red"))  # Set box colors

# This can be plotted in ggplot
# ggplot(data.frame(t(log_counts)), aes(x = condition, y = ENSG00000205426)) +
#   geom_boxplot(fill = c("blue", "red")) +
#   labs(y = "log counts", title = "ENSG00000205426") +
#   theme_classic()

# pdf('ENSG00000205426.pdf', width=5, height=3) # will print a pdf plot (or png)

write.table(res, 'results_dss.csv', sep = ",", col.names = TRUE, row.names = TRUE, quote = FALSE) # to save the DESeq2 object
# dev.off() # to indicate the end of printing to a file and closing of the PDF device.

```
ENSG00000205426 by itself isn't very informative. Look up this gene on [GeneCards](https://www.genecards.org/) or [HGNC](https://genename.org) to find its common name.

To convert ENSEMBL ids to gene symbols run:
```{r gene id conversion}
# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("clusterProfiler")
 if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
 BiocManager::install("clusterProfiler")
library(org.Hs.eg.db)
library(clusterProfiler)
gene_map = data.frame(bitr(geneID = rownames(res), 
                           fromType = 'ENSEMBL', 
                           toType = 'SYMBOL', 
                           OrgDb = org.Hs.eg.db, drop = TRUE))
head(gene_map)
gene_map[gene_map$ENSEMBL == "ENSG00000205426",]
gene_map[gene_map$SYMBOL == "KRT81",]

write.table(gene_map, 'gene_symbol.csv', sep = ",", col.names = TRUE, row.names = TRUE, quote = FALSE) # to save the conversion table
```

The next steps typically involve looking at changes in expression levels of individual genes of interest, or performing pathway or Gene Ontology analysis on the data set to determine the underlying biological processes and their relative changes between the two conditions.   


